Fixes a race condition creating input groups in importer

tinwelint opened the pull request:
Now that input may be processed in parallel, different groups of inputs
which have different groups may request to create new groups concurrently.
Groups class wasn't thread safe since there was previously no requirement
for it.

This commit makes Groups#getOrCreate call synchronized to handle concurrent
invocation. It's typically invoked quite rarely so the locking nature of it
is fine.

cl [tools] Fixed a race condition in `neoj4-import` with ID groups which would have some correct relationships considered invalid and logged into bad.log instead

tinwelint committed:
Fixes a race condition creating input groups in importer

Now that input may be processed in parallel, different groups of inputs
which have different groups may request to create new groups concurrently.
Groups class wasn't thread safe since there was previously no requirement
for it.

This commit makes Groups#getOrCreate call synchronized to handle concurrent
invocation. It's typically invoked quite rarely so the locking nature of it
is fine.

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/InputGroupsDeserializer.java:
```diff
@@ -60,14 +60,15 @@
         this.config = config;
         this.idType = idType;
         this.maxProcessors = maxProcessors;
+        this.previousInputProcessors = processors;
```
Can remove initialisation from declaration if initialised here.

burqen commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/InputGroupsDeserializerTest.java:
```diff
@@ -80,7 +84,52 @@ public void shouldBeAbleToAskForSourceInformationEvenBetweenTwoSources() throws
+                    InputNodeDeserialization deserialization =
+                            new InputNodeDeserialization( header, stream, groups, idType.idsAreExternal() );
+                    return new InputEntityDeserializer<>( header, stream, config.delimiter(),
+                            deserialization, decorator, validator, badCollector );
```
This is the same lamba used in `CsvInput#nodes()` and almost the same as `CsvInput#relationships()`.
I think readability would be improved by extracting them to concrete classes.

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/InputGroupsDeserializer.java:
```diff
@@ -60,14 +60,15 @@
 
     InputGroupsDeserializer( Iterator<DataFactory<ENTITY>> dataFactory, Header.Factory headerFactory,
-            Configuration config, IdType idType, int maxProcessors, DeserializerFactory<ENTITY> factory,
+            Configuration config, IdType idType, int maxProcessors, int processors, DeserializerFactory<ENTITY> factory,
```
See over all comment about processors.

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/ParallelInputEntityDeserializer.java:
```diff
@@ -68,7 +68,7 @@
     @SuppressWarnings( "unchecked" )
     public ParallelInputEntityDeserializer( Data<ENTITY> data, Header.Factory headerFactory, Configuration config,
-            IdType idType, int maxProcessors, DeserializerFactory<ENTITY> factory,
+            IdType idType, int maxProcessors, int processors, DeserializerFactory<ENTITY> factory,
```
See over all comment about processors.

burqen commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/GroupsTest.java:
```diff
@@ -0,0 +1,53 @@
+public class GroupsTest
+{
+    @Test
+    public void shouldHandleConcurrentGetOrCreate() throws Throwable
```
Nice

burqen reviewed (changes requested):
Behaviour change looks good.

I'm not entirely sure about why we need to add processors argument to any of the `Deserializers`. Is it only for test purposes? Seems unrelated to the problem that this PR is addressing.

Otherwise, small clean up comments.

tinwelint commented:
@burqen the change simplified tests. The way these classes are used currently there's really no need to specify initial processor count, currently. At the same time this could be seen as a bug because an invocation on `processors(int)` would have no effect before the first call to `hasNext()`, of course we wouldn't need to add this constructor argument for this, we could have tests call `processors(int)` after construction (which they previously not quite couldn't as easily, with the first `hasNext()` thing), but since I was changing things around here and tests looked unnecessarily complicated I though I'd just add it.

tinwelint committed:
Simplified creation of DeserializerFactory

into common implementations in DeserializerFactories as to deduplicate these
implementations found throught main and test code and make that code more readable
